### DISCUSSION TITLE
How to indicate Managed By

### DISCUSSION THREAD

**Discussion by nomad-cyanide** [netbox-community/netbox#11589] [Q&A] [answered]
Hi

I am looking for a way to document which virtual machine is managed by whatever other virtual machine.

We have many virtual firewalls. Each of these firewalls are managed by another virtual machine, a management-vm. Some of the management-hosts are managing several virtual firewalls (as some customers have more than one), others have just one virtual firewall. We are an ISP.
I would ideally like to have a way to set "managed by" on a firewall or "Manages these device" on the management-vm.

I know I can do this using 2 customs fields, but is there a better way? Perhaps something that link the 2 vm's "dynamically".

Please ask if I need to elaborate further

**Comment by candlerb** [comment, marked as answer]
A single "managed by" custom field sounds like the right model to me. On the firewall, set it to point to the manager VM, since this is a many-to-one relationship.

When viewing the manager VM, I presume you'd want to see the reverse relation: a list of firewalls which is managed by that VM. You can do that creating a custom link to `/virtualization/virtual-machines/?cf_managed_by=NNN` (where NNN is the id of the manager VM).  Or you might be able to do it with a plugin.

I think you should avoid having a second, multi-select custom field pointing in the opposite direction, because that doubles the data entry and means inconsistencies are likely; you'd then need to run a report periodically to hunt out and highlight these inconsistencies.

**Reply by nomad-cyanide** [reply]
Ok, so my ideas are not that far off. :)

I have tried messing around with custom links, but can't quite get what I want. This link gives me a correct list of managed devices /virtualization/virtual-machines/?cf_managed_by={{ object.id }} , but is there a way to make the custom link only appear if cf_managed_by is filled in on another device? 
Something like
if cf_managed_by is empty, don't display link.
Besides that, it would be great to not have the link in the top right corner of the page. Just a nice-to-have. 

Could a custom field of type URL with the correct JSON values perhaps get the job done? I have not tested anything yet, just brainstorming.

**Reply by candlerb** [reply]
> is there a way to make the custom link only appear if cf_managed_by is filled in on another device?

You can make custom links appear/disappear based on making the link text empty with a condition.  Here's one I use to make IPMI buttons appear only if the device has exactly one management interface:

```
# Link text
{% if obj.interfaces.filter(mgmt_only=True).count() == 1 and obj.interfaces.get(mgmt_only=True).ip_addresses.count() == 1 %}Open IPMI{% endif %}

# Link URL
https://{{ obj.interfaces.get(mgmt_only=True).ip_addresses.first().address.ip }}/
```

In your case, it would be a question of finding some Jinja2 query which counts how many other VMs have a cf_managed_by pointing at this object. I'm not sure if it's possible, since "obj" is the object you're currently looking at; AFAIK there is no Django ORM reverse relation created for object custom fields; and the [context](https://github.com/netbox-community/netbox/blob/v3.4.3/netbox/extras/templatetags/custom_links.py#L41-L48) passed to the Jinja2 template doesn't include a way to access the top-level models.

Maybe a new feature is required, e.g. an `obj.custom_has_many('managed_by')` relation.

--------
Aside: according to a comment in the code I linked to, `obj` is deprecated and due to be removed in Netbox 3.5. I see that `object` was added in 3.2 (a5c8bbf79):

```
./version-3.2.md:* [#8684](https://github.com/netbox-community/netbox/issues/8684) - Change custom link template context variable `obj` to `object` (backward-compatible)
```

but I couldn't find any deprecation warning, other than comments in the code.

**Reply by nomad-cyanide** [reply]
Ah, it can take an if-statement. Although your solution would be preferable for my case, it doesn't seem possible, so I was thinking of doing something like if the virtuel device role is "checkpoint-domain-management-server" (the slug for it), then display the link, because all CP managers will have one or more virtual firewalls to manage.

That would be something like this, I guess:

{% if object.role.get(role="checkpoint-domain-management-server") == 1 %} { object } manages these devices:{% endif %}

but I get a "dcim.models.devices.DeviceRole object has no attribute 'get'. I have tried a number of different iterations, but there must be something I just don't understand, as none of them have worked.

**Reply by candlerb** [reply]
It'll be easier than that. Something like `object.device_role.slug == "checkpoint-domain-management-server"`. You can test this out using [nbshell](https://docs.netbox.dev/en/stable/administration/netbox-shell/):

```
>>> d = Device.objects.get(id=1)
>>> d
<Device: nuc1>
>>> d.device_role
<DeviceRole: VM Host>
>>> d.device_role.name
'VM Host'
>>> d.device_role.slug
'vm-host'
>>>
```

Another way would be with a tag.

The `get(...)`, `filter(...)` and `count()` were only required in the queries I showed because `obj.interfaces` is a relation, which returns a queryset (a proxy object in the Django ORM which can be used to build database queries).

EDIT: on a virtual machine it's `role` not `device_role`.

```
>>> v = VirtualMachine.objects.filter(role__isnull=False).first()
>>> v
<VirtualMachine: proxmox1>
>>> v.role
<DeviceRole: VM Host>
>>>
```

**Reply by nomad-cyanide** [reply]
I tried this


#Link Text

{% if object.role.slug == "checkpoint-domain-management-server" %} {{ object }} manages these devices:{% endif %}

and it worked like a charm. Thank you very much!!

**Reply by candlerb** [reply]
Great! Be aware that object.role may be null for a virtual machine, so it may be more robust to write

```
{% if object.role and object.role.slug == "whatever" %}...
```
